File: GW3/LogFileHandling/Query/Tokens/TokenOr.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenOr : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();
            return ((parser.PeekChar() == '|' && parser.PeekChar(1) == '|') || parser.PeekString().ToLower() == "or");
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            if (parser.PeekString().ToLower() == "or")
            {
                parser.NextString();
                return new TokenOr { Value = "||" };
            }
            return new TokenOr { Value = "" + parser.NextChar() + parser.NextChar() };
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenOr.ts
class TokenOr extends Token
{

    public tokenType: TokenType = TokenType.TokenOr;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[
        { suggestion: "or" },
        { suggestion: "||" }
    ];

    public canBeUsed(parser: QueryParser): boolean
    {
        parser.skipSpaces();
        return ((parser.peekChar() == '|' && parser.peekChar(1) == '|') || parser.peekString().toLowerCase() == "or");
    }

    public extract(parser: QueryParser): Token
    {
        parser.skipSpaces();
        var token = new TokenOr();
        if (parser.peekString().toLowerCase() == "or")
        {
            parser.nextString();
            token.value = "||";
            return token;
        }
        token.value = parser.nextChar() + parser.nextChar();
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[]
    {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined')
        {
            switch (nextToken.tokenType)
            {
                case TokenType.TokenName:
                    return [new TokenCompare()];
                case TokenType.TokenOpenParenthesis:
                    return [new TokenName(), new TokenOpenParenthesis()];
            }
        }
        return [];
    }
}